rasa/core/policies/rule_policy.py

```diff
@@ -767,23 +761,26 @@ def train(
         self,
         training_trackers: List[TrackerWithCachedStates],
         domain: Domain,
-        interpreter: NaturalLanguageInterpreter,
         **kwargs: Any,
-    ) -> None:
+    ) -> Resource:
         """Trains the policy on given training trackers.
 
         Args:
             training_trackers: The list of the trackers.
             domain: The domain.
-            interpreter: Interpreter which can be used by the polices for featurization.
+
+        Returns:
+            The resource which can be used to load the trained policy.
         """
+        self._validate_against_domain(domain)
```

Comment:
This is now caught when we start training the component. Before, it was caught before, cf. [here](https://github.com/RasaHQ/rasa/blob/0882404686e085cb6cbc58e369240d8c1c962c62/rasa/core/agent.py#L411). If we want to replicate that behaviour, then I think we should make this check a class method and trigger it during whatever validation runs before any training.

Reply:
I moved it here as it "kind of" replicates the previous behavior (just throws later) and we have the loaded `Domain` at hand. Technically we don't have a loaded domain before running the graph (we could make that an exception but I'd probably wouldn't do so unless we have to). 

The other chance would be to make `validate_against_domain` part of a protocol and then do it as part of [this issue](https://github.com/RasaHQ/rasa/issues/9336). I think this would probably be the easiest and cleanest way - what do you think? (also added a note there about this). If it's okay for you I'd keep the check here for now to preserve the basic functionality of this and then potentially revisit this as part of https://github.com/RasaHQ/rasa/issues/9336

Reply:
sounds good - think what we're also missing is all the "required component" stuff and checks for that, no?

Reply:
Yep, will add it to the recipe validation stuff 👍🏻

Reply:
(you actually already added it there 🎉 )